### PR TITLE
Capture webpage with device frame

### LINKED ISSUE #26: Capture webpage with device frame
ref - https://github.com/sindresorhus/pageres/issues/367
adding deviceFrame when capturing a screenshot.

### PR DESCRIPTION
**Pull request by nsisodiya**
Fixes #26

### CODE REVIEW

File: device-frames/frames-config.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module.exports.framesConfig = {
```
Comment (sindresorhus) [inline device-frames/frames-config.js]:
```suggestion
module.exports = {
```

And the filename should be `config.js`.

File: device-frames/frames-config.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+	'iPhone X': {
+		top: 105,
+		left: 105,
+		width: 1338,
+		height: 2650,
+		frameSrc: 'iphone-x-frame.png'
```
Comment (sindresorhus) [inline device-frames/frames-config.js]:
```suggestion
		image: 'iphone-x.png'
```

File: example.js [hunk 3]
```diff
@@ -5,4 +5,16 @@ const captureWebsite = require('.');
...
+
+	await captureWebsite.file('https://trello.com', 'example_iphone_x.png', {
+		emulateDevice: 'iPhone X'
+	});
+	await captureWebsite.file('https://trello.com', 'example_iphone_x_with_frame.png', {
+		emulateDevice: 'iPhone X',
+		deviceFrame: true
+	});
+	await captureWebsite.file('https://trello.com', 'example_iphone_6_with_frame.png', {
+		emulateDevice: 'iPhone 6',
```
Comment (sindresorhus) [inline example.js]: This would not work as there is no iPhone 6 frame, so why are you showing it?

File: index.d.ts [hunk 4]
```diff
@@ -48,6 +48,12 @@ declare namespace captureWebsite {
 		*/
 		scaleFactor?: number;
 
+		/**
+		Make it look like the screenshot was taken on the specified device with device frame.
+
+		@default false
+		*/
+		deviceFrame?: boolean;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
		deviceFrame?: boolean;

```

File: index.js [hunk 5]
```diff
@@ -5,6 +5,8 @@ const fileUrl = require('file-url');
 const puppeteer = require('puppeteer');
 const devices = require('puppeteer/DeviceDescriptors');
 const toughCookie = require('tough-cookie');
+const sharp = require('sharp');
```
Comment (sindresorhus) [inline index.js]: I'm not a big fan of depending on `sharp` as it's a native Node.js addon, which all kinds of problems.

Instead, why can't we just use Canvas in the renderer to combine the images?

File: index.js [hunk 6]
```diff
@@ -217,13 +220,22 @@ const captureWebsite = async (url, options) => {
 	if (!options._keepAlive) {
 		await browser.close();
 	}
+	// If device frame is true, we need to add device frame here.
 
-	return buffer;
+	const dfConfig = framesConfig[options.emulateDevice];
```
Comment (sindresorhus) [inline index.js]: It should throw as useful if you try to use a device without a frame.

File: index.js [hunk 7]
```diff
@@ -217,13 +220,22 @@ const captureWebsite = async (url, options) => {
...
+				{input: `./device-frames/${dfConfig.frameSrc}`}])
+			.toBuffer();
+	}
+
+	return newBuffer;
 };
 
 module.exports.file = async (url, filePath, options = {}) => {
 	const screenshot = await captureWebsite(url, options);
-
```
Comment (sindresorhus) [inline index.js]: Don't do unrelated changes.

File: readme.md [hunk 8]
```diff
@@ -108,6 +108,13 @@ Scale the webpage `n` times.
 
 The default is what you would get if you captured a normal screenshot on a computer with a retina (High DPI) screen.
 
+##### deviceFrame
+
+Type: `boolean`<br>
+Default: `false`
+
+when true, scrrenshot will be captured with device frame.
```
Comment (sindresorhus) [inline readme.md]: This has typos and generally needs to be better written.

File: example.js (hunk 3 above)
Comment (nsisodiya) [reply example.js]: since there is no frame image for iphone6, It will be equivalent to `deviceFrame: false` !

File: index.js (hunk 5 above)
Comment (nsisodiya) [reply index.js]: In node.js, there is no canvas API. Are you suggesting to use this - https://www.npmjs.com/package/canvas
I can use this too. Whatever you suggest !!

Comment (nsisodiya) [reply index.js]: Are you suggesting the following?
1. Create a local html webpage and open it.
2. The page will load an image(device frame) and the main page in an iframe.
3. Take Screenshot using Canvas API.

Comment (sindresorhus) [reply index.js]: No, using Puppeteer. You already have a browser at your disposal.

Could maybe use https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas

File: example.js (hunk 3 above)
Comment (sindresorhus) [reply example.js]: It shouldn't silently ignore what the user is requesting: https://github.com/sindresorhus/capture-website/pull/27#discussion_r287723192

**Comment by sindresorhus** [comment]
Can you add some more frames? Especially the latest Android Pixel.

File: index.js (hunk 6 above)
Comment (nsisodiya) [reply index.js]: please suggest what changes needed.

File: example.js (hunk 3 above)
Comment (nsisodiya) [reply example.js]: @sindresorhus - Should I add `console.log or console.error` to present error to user?

File: index.js (hunk 6 above)
Comment (sindresorhus) [reply index.js]: I’m confused. I already did.

File: example.js (hunk 3 above)
Comment (sindresorhus) [reply example.js]: No, reusable modules should never use console.log directly. It should throw an error.

**Comment by nsisodiya** [comment]
> Can you add some more frames? Especially the latest Android Pixel.

I have done a "Google Image search" to get the `iPhone X` image. We need transparent png images with exact device frames. Also, there will be an issue with copyright too. We need images which are free to reuse too. Finding such images is difficult. You have a good reach via your Twitter. If you post that you need such device frames, people will contribute and the task will be much easier.

File: example.js (hunk 3 above)
Comment (nsisodiya) [reply example.js]: I have added `throw new Error` on incompatible options.

File: index.js (hunk 6 above)
Comment (nsisodiya) [reply index.js]: Sorry, I am unable to understand what you mean by `It should throw`? If you mean `throw new Error` then It is done now.

Comment (sindresorhus) [reply index.js]: Ah, I'm sorry, I didn't realize I wrote it incorrectly. I was on my phone. I meant to write:

> It should throw a useful error if you try to use a device without a frame.

**Comment by sindresorhus** [comment]
I don't want to tweet about it. I'm sure you can find something.

Some things I found from 30 seconds of googling:
- https://www.sketchappsources.com/free-source/2305-google-pixel-device-mockup-sketch-freebie-resource.html
- https://www.figmafreebies.com/download/google-pixel-device-frames/
- https://applypixels.com/resource/pixel-3-xl

**Comment by sindresorhus** [comment]
@nsisodiya Still interested in finishing this?

**Comment by nsisodiya** [comment]
Hi, sorry for delayed response.
I cannot put device frames for all mobiles.
I did this for iPhone. The coding part is done, only images are left.

**Comment by annymosse** [comment]
@nsisodiya @sindresorhus , honestly waiting this feature , pupeter is a great tool but only for GUI(desktop env) not the servers, sorry i'm not too much expert to help you guys .
Note: the keywords of a devices from the url is :skull: :404 [http://viewportsizes.com/](http://viewportsizes.com/).
